Remove wchartype dependency and is_multi_byte [1/2]

nsoranzo opened the pull request:
[wchartype](http://ginstrom.com/code/wchartype.html) won't make sense in Python3, where all strings are UTF-8 encoded.

See individual commits for details.

xref. #1715

nsoranzo committed:
Fix html sniffer to not return True in case of exception

nsoranzo committed:
Remove duplicated functions

Added in commits 777d6a155c369e49127084b63dee5701da7589f1 and
917d5a72e83741594912578ee14d1df19f5974c9

nsoranzo committed:
Update `get_fileobj()` to use utf-8 encoding in text mode

Also, merge its 3 parameters `gzip_only`, `bz2_only`, `zip_only` into
`compressed_formats` (a list of allowed formats).

As a consequence of the changes in `get_fileobj()`, update:
- `files_diff()`
- `get_file_peek()`, which now determines that a file is binary when a
  `UnicodeDecodeError` exception is raised and doesn't need
  `is_multi_byte` any more
- `iter_headers()` and `get_headers()`, which now return Unicode and don't
  need `is_multi_byte` parameter any more

As a consequence of the changes in `get_file_peek()`, update:
- `set_peek()`, which now doesn't need `is_multi_byte` any more

As a consequence of the changes in `get_headers()`, update:
- `guess_ext` and `is_column_based()`, which now determine that a file is
  binary when a `UnicodeDecodeError` exception is raised and don't need
  `is_multi_byte` any more

As a consequence of the changes to `guess_ext`, update:
- `handle_uploaded_dataset_file() doesn't need `is_multi_byte` any more

Also, remove duplicated calls to `get_file_peek()` in
lib/galaxy/datatypes/molecules.py and lib/galaxy/datatypes/msa.py

The `is_multi_byte` was not removed from the signature of `get_file_peek()`
and `set_peek()` in order to preserve compatibility for ToolShed datatypes,
thanks @jmchilton for the review.

jmchilton commented:
Does ``get_file_peek`` still need to take that is multi-byte argument and just ignore it because of tool shed datatype definitions?

nsoranzo commented on lib/galaxy/datatypes/sniff.py:
```diff
@@ -266,7 +259,7 @@ def is_column_based(fname, sep='\t', skip=0, is_multi_byte=False):
     True
     """
-    headers = get_headers(fname, sep, is_multi_byte=is_multi_byte)
+    headers = get_headers(fname, sep)
```
Need to add a try/except UnicodeDecodeError around this line.

nsoranzo committed:
Add missing newline at end of some test files

It gets added anyway when the file is uploaded to Galaxy due to the
default "Use POSIX standard" (to_posix_lines) option. Many tests which
simply copy them with cat and compare them would fail a strict diff test.

nsoranzo committed:
Improve and test the PDF sniffer

Read just the first 4 bytes in binary mode.

nsoranzo commented:
> Does `get_file_peek` still need to take that is multi-byte argument and just ignore it because of tool shed datatype definitions?

@jmchilton Then I should probably also re-add the `is_multi_byte` param to all the `set_peek()` methods, since some TS datatype subclass may be calling the parent's method when overriding it. What do you think?

nsoranzo committed:
Fix `SnpSiftDbNSFP.generate_primary_file()` to return a string

Also use a `with` statement to open/close a file in
`SnpSiftDbNSFP.regenerate_primary_file()` .

nsoranzo committed:
Add a datatype to sniff_order only if it has a sniff() method

This reduces the number of default sniffers from 412 to 246.

jmchilton commented:
@nsoranzo Sounds fair - sorry about these data types - they are why we can't have nice things.

nsoranzo commented:
@jmchilton Restored the `is_multi_byte` in the signatures.

jmchilton commented:
Thanks for supporting those datatypes and all of this - really nice work!

nsoranzo commented:
Thanks for the review and merge, part 2/2 coming in the next few days, I just have to split changes in commits and test.
